posts/2024-10-17-Rust-1.82.0.md

```diff
@@ -0,0 +1,319 @@
+---
+layout: post
+title: "Announcing Rust 1.82.0"
+author: The Rust Release Team
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.82.0. Rust is a programming language empowering everyone to build reliable and efficient software.
+
+If you have a previous version of Rust installed via `rustup`, you can get 1.82.0 with:
+
+```console
+$ rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`](https://www.rust-lang.org/install.html) from the appropriate page on our website, and check out the [detailed release notes for 1.82.0](https://doc.rust-lang.org/nightly/releases.html#version-1820-2024-10-17).
+
+If you'd like to help us out by testing future releases, you might consider updating locally to use the beta channel (`rustup default beta`) or the nightly channel (`rustup default nightly`). Please [report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you might come across!
+
+## What's in 1.82.0 stable
+
+### `cargo info`
+
+Cargo now has an [`info` subcommand](https://doc.rust-lang.org/nightly/cargo/commands/cargo-info.html) to display information about a package in the registry, fulfilling a [long standing request](https://github.com/rust-lang/cargo/issues/948) just shy of its tenth anniversary! Several third-party extensions like this have been written over the years, and this implementation was developed as [cargo-information](https://crates.io/crates/cargo-information) before merging into Cargo itself.
+
+For example, here's what you could see for `cargo info cc`:
+
+<pre>
+<b><span style="color:#0A0">cc</span></b> <b><span style="color:#0AA">#build-dependencies</span></b>
+A build-time dependency for Cargo build scripts to assist in invoking the native
+C compiler to compile native C code into a static archive to be linked into Rust
+code.
+<b><span style="color:#0A0">version:</span></b> 1.1.23 <b><span style="color:#A50">(latest 1.1.30)</span></b>
+<b><span style="color:#0A0">license:</span></b> MIT OR Apache-2.0
+<b><span style="color:#0A0">rust-version:</span></b> 1.63
+<b><span style="color:#0A0">documentation:</span></b> https://docs.rs/cc
+<b><span style="color:#0A0">homepage:</span></b> https://github.com/rust-lang/cc-rs
+<b><span style="color:#0A0">repository:</span></b> https://github.com/rust-lang/cc-rs
+<b><span style="color:#0A0">crates.io:</span></b> https://crates.io/crates/cc/1.1.23
+<b><span style="color:#0A0">features:</span></b>
+  jobserver = []
+  parallel  = [dep:libc, dep:jobserver]
+<b><span style="color:#0AA">note</span></b><b>:</b> to see how you depend on cc, run `<b><span style="color:#0AA">cargo tree --invert --package cc@1.1.23</span></b>`
+</pre>
+
+By default `cargo info` describes the package version in the local `Cargo.lock`, if any, but as you can see it will indicate when there's a newer version too, and `cargo info cc@1.1.30` would report on that. The feature list also indicates a `+` next to any that are enabled.
+
```

Comment:
````suggestion

### Replacing `Captures` trick with `use<..>` syntax

Rust now supports `use<..>` syntax within certain impl Trait bounds to control which generic lifetime parameters are captured.

Return-position impl Trait (RPIT) types in Rust *capture* certain generic parameters.  Capturing a generic parameter allows that parameter to be used in the hidden type.  That in turn affects borrow checking.

In Rust 2021 and earlier editions, lifetime parameters are not captured in opaque types on bare functions and on functions and methods of inherent impls unless those lifetime parameters are mentioned syntactically in the opaque type.  E.g., this is an error:

```rust
//@ edition: 2021
fn f(x: &()) -> impl Sized { x }
```

```
error[E0700]: hidden type for `impl Sized` captures lifetime that does not appear in bounds
 --> src/main.rs:1:30
  |
1 | fn f(x: &()) -> impl Sized { x }
  |         ---     ----------   ^
  |         |       |
  |         |       opaque type defined here
  |         hidden type `&()` captures the anonymous lifetime defined here
  |
help: add a `use<...>` bound to explicitly capture `'_`
  |
1 | fn f(x: &()) -> impl Sized + use<'_> { x }
  |                            +++++++++
```

With the new `use<..>` syntax, we can fix this, as suggested in the error, by writing:

```rust
fn f(x: &()) -> impl Sized + use<'_> { x }
```

Previously, correctly fixing this class of error required defining a dummy trait, conventionally called `Captures`, and using it as follows:

```rust
trait Captures<T: ?Sized> {}
impl<T: ?Sized, U: ?Sized> Captures<T> for U {}

fn f(x: &()) -> impl Sized + Captures<&'_ ()> { x }
```

That was called "the `Captures` trick", and it was a bit baroque and subtle.  It's no longer needed.

There are some limitations to what is being stabilized.  The `use<..>` syntax cannot currently appear within traits or within trait impls, and it must list all in-scope generic type and const parameters.  We hope to lift these restrictions over time.

Note that in Rust 2024, the example above will "just work" without needing `use<..>` syntax (or any tricks).  This is because in the new edition, opaque types will automatically capture all lifetime parameters in scope.  This is a better default, and we've seen a lot of evidence about how this cleans up code.  In Rust 2024, `use<..>` syntax will serve as an important way of opting-out of that default.

For more details about `use<..>` syntax, capturing, and how this applies to Rust 2024, see the ["RPIT lifetime capture rules"](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/rpit-lifetime-capture.html) chapter of the edition guide.  For details about the overall direction, see our recent blog post, ["Changes to `impl Trait` in Rust 2024"](2024-09-05-impl-trait-capture-rules.md).

````

Reply:
Apologies in advance about the length; this is as short as I could make it.

Reply:
Rust 1.82 is going to be an unusually large release in terms of user-facing things, so it probably makes sense it will be a longer-than-usual blog post.

Reply:
I think the problem here, starting from the section title, is that the "`Captures` trick" is (per RFC 3498) "not widely known, and its purpose is not commonly well understood." So I think a title like "Precise capturing with `use<>` syntax" would be more accessible, and it also needs to motivate why I shouldn't just use `+ '_`, **especially** because that's what previous versions recommended for that same example!

AIUI, the problem with outlives is that it may force an unnecessary bound on type parameters, like `T: 'a`, whereas `use<'a, T>` lets those remain independent. Is that right?

Reply:
Yes, I think in an earlier draft I'd titled the section...

```markdown
### Precise capturing `use<..>` syntax
```

...and that'd still be good.

I'll write something up about the `+ '_` bit.  I'd tried to avoid it here, as it gets somewhat detailed, but you're right about why it should be said.

Reply:
Ok, I've shuffled everything else around, but I left this out for now to await your update.

Thanks a lot for your help!

Reply:
BTW, I don't think the relative `md` link for the Changes post will work in the site generator, so that should use the published URL.

Reply:
I incorporated the title change and mentioned the old compiler suggestion.  I'll try to make some other changes late too, but if not.... this would probably be OK given the links to RFC 3498 where this is discussed further.

> AIUI, the problem with outlives is that it may force an unnecessary bound on type parameters, like `T: 'a`, whereas `use<'a, T>` lets those remain independent. Is that right?

Yes... specifically think about invariant lifetimes where this ends up requiring the lifetimes be exactly equal, resulting in no good way for the caller to provide that.

Reply:
I've now cleaned this up further, added an example of how `+ '_` doesn't work in general inspired by a real case in the compiler, and pushed this to the branch.